### PR TITLE
own realisation of calcualator

### PR DESCRIPTION
**Pull request by Alekearl**

### CODE REVIEW

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    private Random random = new Random();
+    private ArithmeticalOperator arithmeticalOperator = new Calculator();
+    private ArithmeticOperators[] arrayEnum;
+
+    @Test
+    void randomTest() {
+        arrayEnum = ArithmeticOperators.values();
+        int operator = random.nextInt(5);
+        firstNumber = random.nextDouble() * MAX_DOUBLE;
+        secondNumber = random.nextDouble() * MAX_DOUBLE;
```
Comment (engineer027) [inline src/test/java/core/basesyntax/CalculatorTest.java:23]: It is bad practice to use random

Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:23]: I understand that, and thank you for remind. But this random method covers moments with different solutions(^__^). So, I would like to leave it here.

**Review by engineer027 — APPROVED** [review]

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]
Tests are good but let's redo your calculator using char operator(and tests according to that)

File: src/main/java/core/basesyntax/Calculator.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package core.basesyntax;
+
+public class Calculator implements ArithmeticalOperator {
+
+    @Override
+    public double calculate(ArithmeticOperators operations, double firstValue, double secondValue) {
```
[the remark this answers is not in the document]
Comment (Alekearl) [reply src/main/java/core/basesyntax/Calculator.java]: Changed. Please, check. =)

File: src/main/java/core/basesyntax/ArithmeticOperators.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package core.basesyntax;
+
+public enum ArithmeticOperators {
+    ADDITION('+'),
+    SUBTRACTION('-'),
+    DIVISION('/'),
+    MULTIPLICATION('*'),
+    RISING_TO_POWER('^');
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArithmeticOperators.java:8]: I think that Dmytro meant to ask you to remove this enum and simply pass chars to calculate() method in tests

File: src/main/java/core/basesyntax/ArithmeticalOperator.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package core.basesyntax;
+
+public interface ArithmeticalOperator {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArithmeticalOperator.java:3]:
```suggestion
public interface Calculable {
```

File: src/main/java/core/basesyntax/ArithmeticalOperator.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArithmeticalOperator.java:4]: redundant empty line

File: src/main/java/core/basesyntax/ArithmeticalOperator.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+    double calculate(Character operations, double firstValue, double secondValue);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArithmeticalOperator.java:5]:
```suggestion
    double calculate(char operation, double firstValue, double secondValue);
```

File: src/main/java/core/basesyntax/Calculator.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package core.basesyntax;
+
+public class Calculator implements ArithmeticalOperator {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Calculator.java:4]: redundant empty line

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
+package core.basesyntax;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import java.util.Random;
+import org.junit.jupiter.api.Test;
+
+class CalculatorTest {
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:9]: please add access modifiers to the class and all the methods

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    private static final double MAX_DOUBLE = Double.MAX_EXPONENT;
+    private static final double MIN_DOUBLE = Double.MIN_EXPONENT;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:11]: check these values - these are not min/max double values
let's rename them if you use them as common positive and negative values

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    private double firstNumber;
+    private double secondNumber;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:13]: please make these variables local

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    private Random random = new Random();
+    private ArithmeticalOperator arithmeticalOperator = new Calculator();
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:15]: let’s create a Calculator instance in a method that is run before all tests

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
for +, -, / and * let's also add tests for edge values (Double.MIN/MAX_VALUE)

File: src/main/java/core/basesyntax/ArithmeticOperators.java (hunk 3 above)
Comment (Alekearl) [reply src/main/java/core/basesyntax/ArithmeticOperators.java:8]: I joined Dmytro suggestion to my code and neither want to remove my part of code. But if you insist, I'll refactor it. ^__^

File: src/main/java/core/basesyntax/ArithmeticalOperator.java (hunk 4 above)
Comment (Alekearl) [reply src/main/java/core/basesyntax/ArithmeticalOperator.java]: Changed

File: src/main/java/core/basesyntax/ArithmeticalOperator.java (hunk 5 above)
Comment (Alekearl) [reply src/main/java/core/basesyntax/ArithmeticalOperator.java:4]: Done.

File: src/main/java/core/basesyntax/Calculator.java (hunk 7 above)
Comment (Alekearl) [reply src/main/java/core/basesyntax/Calculator.java:4]: Removed

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 8 above)
Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:9]: changed

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 10 above)
Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:13]: done

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 11 above)
Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:15]: changed

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 9 above)
Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:11]: changed

File: src/main/java/core/basesyntax/ArithmeticOperators.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+    private Character character;
+
+    ArithmeticOperators(Character character) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArithmeticOperators.java:12]:
```suggestion
    public ArithmeticOperators(Character character) {
```

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class CalculatorTest {
+    private static final double MAX_DOUBLE = Double.MAX_VALUE;
+    private static final double MIN_DOUBLE = Double.MIN_VALUE;
+    private Random random;
+    private Calculable calculable;
+    private ArithmeticOperators[] operations;
+    private ArithmeticOperators operator;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:16]: `ArithmeticOperators operator;` - let's make this variable local in each test

File: src/test/java/core/basesyntax/CalculatorTest.java [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+
+    @BeforeEach
+    public void setUp() {
+        random = new Random();
+        calculable = new Calculator();
+        operations = ArithmeticOperators.values();
+    }
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/CalculatorTest.java:23]: currently we are initializing the fields before each test - can we optimize it? see checklist

**Review by KseniiaMakarova — APPROVED** [review]

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 14 above)
Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:23]: changed. Have a nice weekend. ;)

File: src/test/java/core/basesyntax/CalculatorTest.java (hunk 13 above)
Comment (Alekearl) [reply src/test/java/core/basesyntax/CalculatorTest.java:16]: relocated.

File: src/main/java/core/basesyntax/ArithmeticOperators.java (hunk 12 above)
Comment (Alekearl) [reply src/main/java/core/basesyntax/ArithmeticOperators.java:12]: Modifier public not allowed here.
